### ISSUE TITLE
Getting 404 error for .getPageSource() and .getWindowRect() commands after update to xCode 15

### ISSUE THREAD

**Issue by DanKovt** [appium/appium#19566]
### Do I have the most recent component updates?

- [X] I use the most recent available driver/plugin and server versions

### Is the component officially supported by the Appium team?

- [X] I have verified the component repository is present under the Appium organization in GitHub

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Our developers updated code to build our RN application using xCode 15 and after that our tests stopped working. 

Session capabilities: 

[XCUITestDriver@5f05] Creating session with W3C capabilities: {
[XCUITestDriver@5f05]   "alwaysMatch": {
[XCUITestDriver@5f05]     "platformName": "iOS",
[XCUITestDriver@5f05]     "appium:deviceName": "iPhone 12",
[XCUITestDriver@5f05]     "appium:platformVersion": "17",
[XCUITestDriver@5f05]     "appium:wdaLocalPort": "8213 ",
[XCUITestDriver@5f05]     "appium:udid": "xxxxxxxxxxxxxxxxxxx",
[XCUITestDriver@5f05]     "appium:newCommandTimeout": 1800,
[XCUITestDriver@5f05]     "appium:noReset": true,
[XCUITestDriver@5f05]     "appium:automationName": "XCUITest",
[XCUITestDriver@5f05]     "appium:bundleId": "com.app.bundleId",
[XCUITestDriver@5f05]     "appium:wdaLaunchTimeout": 80000,
[XCUITestDriver@5f05]     "appium:disableIdLocatorAutocompletion": true,
[XCUITestDriver@5f05]     "appium:showXcodeLog": true,
[XCUITestDriver@5f05]     "appium:waitForIdleTimeout": "0",
[XCUITestDriver@5f05]     "appium:includeSafariInWebviews": true,
[XCUITestDriver@5f05]     "appium:connectHardwareKeyboard": true
[XCUITestDriver@5f05]   },
[XCUITestDriver@5f05]   "firstMatch": [
[XCUITestDriver@5f05]     {}
[XCUITestDriver@5f05]   ]
[XCUITestDriver@5f05] }

Appium logs: 

[XCUITestDriver@5f05 (e6e11861)] Calling AppiumDriver.getWindowRect() with args: ["e6e11861-3145-4f96-9237-67df98418466"]
[XCUITestDriver@5f05 (e6e11861)] Executing command 'getWindowRect'
[XCUITestDriver@5f05 (e6e11861)] Proxying [GET /window/size] to [GET http://127.0.0.1:8213/session/A32E4A1E-4EE2-403C-9D6D-066D9871D8B6/window/size] with no body
[Xcode] 2023-12-21 01:08:42.684968+0200 WebDriverAgentRunner-Runner[11307:9199318] Getting the most recent active application (out of 1 total items)
[Xcode] 
[Xcode]     t =     3.40s Requesting snapshot of accessibility hierarchy for app with pid 11299
[Xcode] 
[Xcode]     t =     4.42s Requesting snapshot of accessibility hierarchy for app with pid 11299
[Xcode] 
[Xcode]     t =     5.44s Requesting snapshot of accessibility hierarchy for app with pid 11299
[Xcode] 
[XCUITestDriver@5f05 (e6e11861)] Got response with status 404: {"value":{"error":"stale element reference","message":"The previously found element \"Application 'com.app.bundleId'\" is not present in the current view anymore. Make sure the application UI has the expected state. Original error: Error getting main window kAXErrorServerNotFound","traceback":"(\n\t0   CoreFoundation                      0x00000001a1ef087c B37A1273-B0C2-312E-9E0B-7143FB645C69 + 968828\n\t1   libobjc.A.dylib                     0x000000019a20fc00 objc_exception_throw + 60\n\t2   WebDriverAgentLib                   0x0000000107399cc0 -[XCUIElement(FBUtilities) fb_takeSnapshot] + 756\n\t3   WebDriverAgentLib                   0x00000001073913fc -[XCUIElement(WebDriverAttributesForwarding) fb_snapshotForAttributeName:] + 656\n\t4   WebDriverAgentLib                   0x0000000107391634 -[XCUIElement(WebDriverAttributesForwarding) forwardingTargetForSelector:] + 220\n\t5   CoreFoundation                      0x00000001a1e35ab0 B37A1273-B0C2-312E-9E0B-7143FB645C69 + 203440\n\t6   CoreFoundat...
[W3C] Matched W3C error code 'stale element reference' to StaleElementReferenceError
[XCUITestDriver@5f05 (e6e11861)] Encountered internal error running command: StaleElementReferenceError: The previously found element "Application 'com.app.bundleId'" is not present in the current view anymore. Make sure the application UI has the expected state. Original error: Error getting main window kAXErrorServerNotFound
[XCUITestDriver@5f05 (e6e11861)]     at errorFromW3CJsonCode (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/protocol/errors.js:1064:25)
[XCUITestDriver@5f05 (e6e11861)]     at ProxyRequestError.getActualError (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/protocol/errors.js:933:14)
[XCUITestDriver@5f05 (e6e11861)]     at JWProxy.command (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/jsonwp-proxy/proxy.js:353:19)
[XCUITestDriver@5f05 (e6e11861)]     at runMicrotasks (<anonymous>)
[XCUITestDriver@5f05 (e6e11861)]     at processTicksAndRejections (internal/process/task_queues.js:95:5)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.proxyCommand (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/proxy-helper.js:109:35)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.getWindowSizeNative (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/general.js:278:12)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.getWindowSize (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/general.js:52:14)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.getWindowRect (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/general.js:121:29)
[HTTP] <-- GET /wd/hub/session/e6e11861-3145-4f96-9237-67df98418466/window/rect 404 2285 ms - 1706
[HTTP] 
[HTTP] --> GET /wd/hub/session/e6e11861-3145-4f96-9237-67df98418466/source
[HTTP] {}
[XCUITestDriver@5f05 (e6e11861)] Calling AppiumDriver.getPageSource() with args: ["e6e11861-3145-4f96-9237-67df98418466"]
[XCUITestDriver@5f05 (e6e11861)] Executing command 'getPageSource'
[XCUITestDriver@5f05 (e6e11861)] Matched '/source?format=xml&scope=AppiumAUT&excluded_attributes=accessible%2Cenabled%2Cvisible' to command name 'getPageSource'
[XCUITestDriver@5f05 (e6e11861)] Proxying [GET /source?format=xml&scope=AppiumAUT&excluded_attributes=accessible%2Cenabled%2Cvisible] to [GET http://127.0.0.1:8213/session/A32E4A1E-4EE2-403C-9D6D-066D9871D8B6/source?format=xml&scope=AppiumAUT&excluded_attributes=accessible%2Cenabled%2Cvisible] with no body
[Xcode] 2023-12-21 01:08:44.925142+0200 WebDriverAgentRunner-Runner[11307:9199318] Getting the most recent active application (out of 1 total items)
[Xcode] 
[Xcode] 2023-12-21 01:08:44.933210+0200 WebDriverAgentRunner-Runner[11307:9199318] The following attributes were requested to be included into the XML: {(
[Xcode]     FBYAttribute,
[Xcode]     FBLabelAttribute,
[Xcode]     FBIndexAttribute,
[Xcode]     FBXAttribute,
[Xcode]     FBNameAttribute,
[Xcode]     FBHeightAttribute,
[Xcode]     FBValueAttribute,
[Xcode]     FBWidthAttribute,
[Xcode]     FBTypeAttribute
[Xcode] )}
[Xcode]     t =     5.58s Requesting snapshot of accessibility hierarchy for app with pid 11299
[Xcode] 
[Xcode]     t =     6.59s Requesting snapshot of accessibility hierarchy for app with pid 11299
[Xcode] 
[XCUITestDriver@5f05 (e6e11861)] Got response with status 404: {"value":{"error":"stale element reference","message":"The previously found element \"Application 'com.app.bundleId'\" is not present in the current view anymore. Make sure the application UI has the expected state. Original error: Error getting main window kAXErrorServerNotFound","traceback":"(\n\t0   CoreFoundation                      0x00000001a1ef087c B37A1273-B0C2-312E-9E0B-7143FB645C69 + 968828\n\t1   libobjc.A.dylib                     0x000000019a20fc00 objc_exception_throw + 60\n\t2   WebDriverAgentLib                   0x0000000107399cc0 -[XCUIElement(FBUtilities) fb_takeSnapshot] + 756\n\t3   WebDriverAgentLib                   0x0000000107399f04 -[XCUIElement(FBUtilities) fb_snapshotWithAttributes:maxDepth:] + 152\n\t4   WebDriverAgentLib                   0x0000000107350b14 +[FBXPath writeXmlWithRootElement:indexPath:elementStore:includedAttributes:writer:] + 708\n\t5   WebDriverAgentLib                   0x0000000107350288 +[FBXPath xmlRepresentationWithRootElement:writer:elementStore:qu...
[W3C] Matched W3C error code 'stale element reference' to StaleElementReferenceError
[XCUITestDriver@5f05 (e6e11861)] Encountered internal error running command: StaleElementReferenceError: The previously found element "Application 'com.app.bundleId" is not present in the current view anymore. Make sure the application UI has the expected state. Original error: Error getting main window kAXErrorServerNotFound
[XCUITestDriver@5f05 (e6e11861)]     at errorFromW3CJsonCode (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/protocol/errors.js:1064:25)
[XCUITestDriver@5f05 (e6e11861)]     at ProxyRequestError.getActualError (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/protocol/errors.js:933:14)
[XCUITestDriver@5f05 (e6e11861)]     at JWProxy.command (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/jsonwp-proxy/proxy.js:353:19)
[XCUITestDriver@5f05 (e6e11861)]     at runMicrotasks (<anonymous>)
[XCUITestDriver@5f05 (e6e11861)]     at processTicksAndRejections (internal/process/task_queues.js:95:5)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.proxyCommand (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/proxy-helper.js:109:35)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.mobileGetSource (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/source.js:50:35)
[XCUITestDriver@5f05 (e6e11861)]     at XCUITestDriver.getPageSource (/Users/daniil.kovtun/.appium/node_modules/appium-xcuitest-driver/lib/commands/source.js:24:9)
[HTTP] <-- GET /wd/hub/session/e6e11861-3145-4f96-9237-67df98418466/source 404 2119 ms - 1563
[HTTP] 
[Xcode]     t =     7.61s Requesting snapshot of accessibility hierarchy for app with pid 11299
[Xcode] 
[HTTP] --> GET /wd/hub/session/e6e11861-3145-4f96-9237-67df98418466/timeouts
[HTTP] {}
[XCUITestDriver@5f05 (e6e11861)] Calling AppiumDriver.getTimeouts() with args: ["e6e11861-3145-4f96-9237-67df98418466"]
[XCUITestDriver@5f05 (e6e11861)] Executing command 'getTimeouts'
[XCUITestDriver@5f05 (e6e11861)] Responding to client with driver.getTimeouts() result: {"command":1800000,"implicit":0}
[HTTP] <-- GET /wd/hub/session/e6e11861-3145-4f96-9237-67df98418466/timeouts 200 10 ms - 42

### Expected Behavior

We should be able to get page source for RN apps built with xCode 15

### Minimal Reproducible Example

Cant provide

### Environment

- Operating system: Mac OS Sonoma 14.1.1
- Appium server version (output of `appium --version`): 2.2.3
- Appium driver(s) and their version(s): xcuitest@5.12.1
- Appium plugin(s) and their version(s): none
- Node.js version (output of `node --version`): v14.20.1
- `npm` version (output of `npm --version`): 9.3.0
- Last component(s) version which did _not_ exhibit the problem:
- Platform and version under test: iOS 17.1.1
- Real device or emulator/simulator: real device

**Comment by mykola-mokhnach** [comment]
> Error getting main window kAXErrorServerNotFound

Unfortunately this error is coming from xctest itself and we cannot do anything to change that behaviour. Try to google it, maybe there are some good advices for possible workarounds [elsewhere](https://github.com/appium/appium/issues?q=is%3Aissue+%22Error+getting+main+window+kAXErrorServerNotFound%22+is%3Aclosed).

**Comment by DanKovt** [comment]
@mykola-mokhnach I've seen all these issues, but maybe you can tell more about problem itself, so we at lease can fix it from our side please? Because we can't understand what exactly cause this issue.

**Comment by mykola-mokhnach** [comment]
Unfortunately I also don't know exactly. The exception is thrown by xctest when it tries to take a snapshot of the app accessibility tree. There must be something preventing it from doing that. Apple did not make it easy - XCTest is a closed-source framework, so only their engineers might know more details.
I would also try to look for device logs, maybe xctest leaves some useful traces there.

**Comment by mykola-mokhnach** [comment]
@Dan-Maor Do you have any ideas on the above?

**Comment by Dan-Maor** [comment]
This sounds similar to an issue I encountered before where `testmanagerd` fails to write automation information to the `standardUserDefaults` because the daemon responsible for the user defaults hanged. This is usually caused by stressing devices for long periods of time by repeatedly uninstalling, reinstalling and starting an automation session against the freshly installed application. This does not happen with every app and the root cause was very challenging to understand at the time. In any case, the only fully valid solution for it was a full device reboot when the issue occurs.

@DanKovt in order to confirm, will you be able to provide a full syslog of a full session creation from a device in the faulty state?

**Comment by DanKovt** [comment]
@Dan-Maor can you provide instructions how I can do this please?

**Comment by Dan-Maor** [comment]
@DanKovt You can either use the the MacOS Console app and retrieve the device log by selecting it from the left pane and then click on "Start Streaming", or by using the `idevicesyslog` utility from libimobiledevice.

**Comment by DanKovt** [comment]
@Dan-Maor Attaching logs for my iPhone from MacOS Console as you requested.
[iPhone logs.pdf](https://github.com/appium/appium/files/13776108/iPhone.logs.pdf)

**Comment by Dan-Maor** [comment]
Thanks @DanKovt . While the log does not show the entire session creation sequence (possibly filtered to only show error level), the first line indicates that this is the same scenario I’ve encountered.

`Not updating lastKnownShmemState in CFPrefsPlistSource<0x6f6c0a440> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: Yes): 113 -> 406`

`testmanagerd` is unable to write a preference value, suggesting that `cfprefsd` might be in a hanged state, and thus cannot write the relevant properties to the `standardUserDefaults` of the application in order to enable automation access.

Like I mentioned previously, the only solution we were able to find was to reboot the device in question when it gets to such a state. Perhaps a factory reset could help as well, but I cannot guarantee that it will resolve the problem permanently.

**Comment by mykola-mokhnach** [comment]
Closed as third party issue

**Comment by DanKovt** [comment]
Hi guys (@Dan-Maor @mykola-mokhnach), our dev team figured out what caused the issue! . You were right, the issue lies with an third library used in our project.

Since we work with react-native, we sometimes need components written in native code that are bridged into JS. This library allowed us to avoid writing bridge classes in Objective-C, enabling us to directly implement such components in Swift. Everything stopped working due to adapting the project for building with Xcode 15.

In short, the problem was caused by use of the class_getClassMethod method. By switching from it to class_copyMethodList, everything started working again 🎉🎉

**Comment by mykola-mokhnach** [comment]
Thanks for the update. I hope it would be useful to other people facing similar issues

**Comment by benzman81** [comment]
> In short, the problem was caused by use of the class_getClassMethod method. By switching from it to class_copyMethodList, everything started working again 🎉🎉

Hi, we also have the problem in a react native app. What 3rd party library was it and how did you find out?
